File: algorithms/java/sorting_algorithms/merge_sort.java
package sorting_algorithms;
import java.util.ArrayList;

public class merge_sort {

	public static void main(String args[]) {

		ArrayList unsortedList = new ArrayList();
		for(int i = 0 ; i < 10 ;i++) {
			unsortedList.add(createRondomNum(100));
		}
		System.out.println("The unsorted list : "+unsortedList);
		merge_sort(unsortedList);
		System.out.println("The sorted list : "+unsortedList);
	}
	
	private static void merge_sort(ArrayList list) {
		if(list.size()>1) {
			int x = (int)list.size()/2;
			
			ArrayList leftList = new ArrayList();
			for(int i = 0 ;i < x;i++) {
				leftList.add(list.get(i));
			}
			//System.out.println(leftList);
			
			ArrayList rightList = new ArrayList();
			for(int i = x ;i < list.size();i++) {
				rightList.add(list.get(i));
			}
			//System.out.println(rightList);
			
			merge_sort(leftList);
			merge_sort(rightList);
			
			int indexLeft = 0;
			int indexRight = 0;
			int indexMain = 0;
			
			while(indexLeft < leftList.size() & indexRight < rightList.size()) {
				if ((int)leftList.get(indexLeft) < (int)rightList.get(indexRight)) {
					list.set(indexMain, (int)leftList.get(indexLeft));
					indexLeft++;
				}else {
					list.set(indexMain, (int)rightList.get(indexRight));
					indexRight++;
				}
				indexMain++;
			}
			
			while(indexLeft < leftList.size()) {
				list.set(indexMain, (int)leftList.get(indexLeft));
				indexLeft++;
				indexMain++;
			}
			while(indexRight < rightList.size()) {
				list.set(indexMain, (int)rightList.get(indexRight) );
				indexRight++;
				indexMain++;
			}
		}
	}
	private static int createRondomNum(int i) {
		int number = (int) ( i*Math.random()) ;
		return number;
	}
}

File: algorithms/python/sorting_algorithms/merge_sort.py
import  random
def merge_sort(_list):
    if len(_list) > 1:

        x = len(_list) // 2
        left_list = _list[:x]
        right_list = _list[x:]
        merge_sort(left_list)
        merge_sort(right_list)

        indexLeft=0
        indexRight=0
        indexMain = 0

        while indexLeft < len(left_list) and indexRight < len(right_list):
            if left_list[indexLeft] < right_list[indexRight]:
                _list[indexMain] = left_list[indexLeft]
                indexLeft += 1
            else:
                _list[indexMain] = right_list[indexRight]
                indexRight += 1
            indexMain += 1

        while indexLeft < len(left_list):
            _list[indexMain] = left_list[indexLeft]
            indexLeft += 1
            indexMain += 1

        while indexRight < len(right_list):
            _list[indexMain] = right_list[indexRight]
            indexRight += 1
            indexMain += 1

unsortedList = []
for i in range(10):
    unsortedList.append(random.randint(0,100))
print(f'The unsorted list: {unsortedList}')
merge_sort(unsortedList)
print(f'The sorted list: {unsortedList}')